### PR TITLE
RUMM-1679 Compress HTTP body using `deflate` (ETF RFC 1950)

### PR DESCRIPTION
**Pull request by maxep**
### What and why?

Apply HTTP body compression using `deflate` data format as described in [IETF RFC 1950](https://www.ietf.org/rfc/rfc1950.txt)

### How?

Using the `zlib` structure (defined in [IETF RFC 1950](https://www.ietf.org/rfc/rfc1950.txt)) with the `DEFLATE` compression algorithm (defined in [IETF RFC 1951](https://www.ietf.org/rfc/rfc1951.txt)).

```
+  Header   +     Raw DEFLATE     +   Checksum    +
+-----+-----+=====================+---+---+---+---+
| CMF | FLG |...compressed data...|    ADLER32    |
+-----+-----+=====================+---+---+---+---+
```

The `DEFLATE` implementation uses [`compression_encode_buffer(_:_:_:_:_:_:)`](https://developer.apple.com/documentation/compression/1480986-compression_encode_buffer) from the [`Compression`](https://developer.apple.com/documentation/compression) framework by allocating a destination buffer of source size and copying the result into a `Data` structure. In the worst possible case, where the compression expands the data size, the destination buffer becomes too small and deflation returns `nil`.

The `Adler32` checksum is performed by using the [`adler32(_:_:_)`](https://developer.apple.com/documentation/kernel/1546941-adler32/) from `zlib` library.

After successful compression, `RequestBuilder` will set the request's `httpBody` with the deflated data and set the `Content-Encoding` header value to `deflate`. In case of compression failure, the `httpBody` will fallback to the uncompressed data.

### Tests

`ServerMock` is now provided with `unzip` and `inflate` methods to automatically decompress `httpBody` during tests.

> I'm on the fence of moving decompression capabilities to `http-server-mock`, but that would prevent from reusing those methods in `DataCompressionTests`. 

### Review checklist

- [x] Feature or bugfix MUST have appropriate tests (unit, integration)
- [x] Make sure each commit and the PR mention the Issue number or JIRA reference

### CODE REVIEW

File: Sources/Datadog/Core/Upload/DataCompression.swift [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+///
+/// - Parameter data: Source data to deflate
+/// - Returns:  The compressed data format.
+internal func zip(_ data: Data) -> Data? {
+    // 2 bytes header - defines the compression mode
+    // +---+---+
+    // |CMF|FLG|
+    // +---+---+
+    // ref. https://datatracker.ietf.org/doc/html/rfc1950#section-2.2
+    let header = Data([0x78, 0x5e])
```
Comment (mariusc83) [inline Sources/Datadog/Core/Upload/DataCompression.swift:27]: are these flags randomly generated ?

File: Sources/Datadog/Core/Upload/DataCompression.swift [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        // The Adler-32 checksum should be initialized to 1 as described in
+        // https://datatracker.ietf.org/doc/html/rfc1950#section-8
+        return zlib.adler32(1, ptr, uInt(data.count))
+    }
+
+    guard let checksum = adler else {
+        return nil
+    }
+
+    var bytes = UInt32(checksum).bigEndian
```
Comment (mariusc83) [inline Sources/Datadog/Core/Upload/DataCompression.swift:98]: Why is this needed ? is `adler32` using `littleEndian` ?

Comment (maxep) [reply Sources/Datadog/Core/Upload/DataCompression.swift:98]: ARM is most likely little-endian. In anyways, `.bigEndian` will only reverse bytes if necessary and this is needed to convert this UInt32 into a bytes array for the network byte order.

File: Sources/Datadog/Core/Upload/DataCompression.swift (hunk 1 above)
Comment (maxep) [reply Sources/Datadog/Core/Upload/DataCompression.swift:27]: This is coming from [mw99/DataCompression](https://github.com/mw99/DataCompression/blob/master/Sources/DataCompression/DataCompression.swift#L98) which is using the same algorithm defined in the Apple's `Compression` framework: [COMPRESSION_ZLIB](https://developer.apple.com/documentation/compression/compression_zlib)

**Review by mariusc83 — APPROVED** [review]
LGTM but I am not the expert here :)

**Review by ncreated — COMMENTED** [review]
Great job, very well explained and super clear 👌. I left few comments, with the test `Data` randomness being the most important one.

Out of curiosity, do we have any benchmarks on the compression stuff? What could be the estimated time impact on building the request?

File: Tests/DatadogTests/Datadog/Core/Upload/DataCompressionTests.swift [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import XCTest
+@testable import Datadog
+
+class DataCompressionTests: XCTestCase {
+    func testFuzzy_Adler32() {
```
[the remark this answers is not in the document]
Comment (maxep) [reply Tests/DatadogTests/Datadog/Core/Upload/DataCompressionTests.swift:11]: Yes, I just need to get used to it :)

File: Tests/DatadogTests/Datadog/Mocks/SystemFrameworks/FoundationMocks.swift [hunk 4]
```diff
@@ -50,12 +50,16 @@ extension Data: AnyMockable, RandomMockable {
         return Data(repeating: byte, count: count)
     }
 
-    static func mock(ofSize size: UInt64) -> Data {
+    static func mock<Size>(ofSize size: Size) -> Data where Size: BinaryInteger {
         return mockRepeating(byte: 0x41, times: Int(size))
     }
 
+    static func mockRandom<Size>(ofSize size: Size) -> Data where Size: BinaryInteger {
+        return mockRepeating(byte: .random(in: 0x00...0xFF), times: Int(size))
```
[the remark this answers is not in the document]
Comment (maxep) [reply Tests/DatadogTests/Datadog/Mocks/SystemFrameworks/FoundationMocks.swift:58]: Definitely, good catch!

File: Sources/Datadog/Core/Upload/DataCompression.swift [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+///
+/// The `Compression` library implements the zlib encoder at level 5 only. This compression level
+/// provides a good balance between compression speed and compression ratio.
+///
+/// The encoded format is the ZLIB Compressed Data Format as described in IETF RFC 1950
+/// https://datatracker.ietf.org/doc/html/rfc1950
+///
+/// - Parameter data: Source data to deflate
+/// - Returns:  The compressed data format.
+internal func zip(_ data: Data) -> Data? {
```
[the remark this answers is not in the document]
Comment (maxep) [reply Sources/Datadog/Core/Upload/DataCompression.swift:21]: zip sequences... 🙄 
I will try to find a better name!

File: Sources/Datadog/Core/Upload/RequestBuilder.swift [hunk 6]
```diff
@@ -126,9 +127,15 @@ internal struct RequestBuilder {
...
 
         request.httpMethod = "POST"
-        request.allHTTPHeaderFields = headers
-        request.httpBody = data
 
+        if let body = zip(data) {
+            headers[HTTPHeader.contentEncodingHeaderField] = "deflate"
+            request.httpBody = body
+        } else {
+            request.httpBody = data
```
[the remark this answers is not in the document]
Comment (maxep) [reply Sources/Datadog/Core/Upload/RequestBuilder.swift:135]: That's a good point! we could add an internal `warning` during our next beta? to see when payloads are not compressed and for what data size. But I would remove it for the release as this won't be actionable as you said.
I will also add another test to `RequestBuilderTests` which will create requests with an increasing size of payload and expect compression.

File: Tests/DatadogTests/Datadog/Mocks/SystemFrameworks/FoundationMocks.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (maxep) [reply Tests/DatadogTests/Datadog/Mocks/SystemFrameworks/FoundationMocks.swift:58]: @ncreated, I've been adding a better randomisation using [SecRandomCopyBytes](https://developer.apple.com/documentation/security/1399291-secrandomcopybytes) BUT all compression started failing.. So I've been reading more about the subject and discovered that random data (especially cryptographically strong random numbers) **cannot** be compressed. In [this wikipedia article](https://en.wikipedia.org/wiki/Lossless_compression#Lossless_compression_techniques) you can read: 
> The primary encoding algorithms used to produce bit sequences are Huffman coding (also used by the deflate algorithm) and arithmetic coding. Arithmetic coding achieves compression rates close to the best possible for a particular statistical model, which is given by the information entropy, whereas Huffman compression is simpler and faster but produces poor results for models that deal with symbol probabilities close to 1.

> In particular, files of random data cannot be consistently compressed by any conceivable lossless data compression algorithm; indeed, this result is used to define the concept of randomness in Kolmogorov complexity.[18]
> It is probably impossible to create an algorithm that can losslessly compress any data

In conclusion, a good randomised array of bytes won't be compressible.

[the remark this answers is not in the document]
Comment (buranmert) [reply Tests/DatadogTests/Datadog/Mocks/SystemFrameworks/FoundationMocks.swift]: we can create such random strings by selecting `randomElement()`s from a pool of keywords
along these lines:
```swift
let keywords = ["foo", "bar", ...]
let randomString = (1...100).map { _ in keywords.randomElement()! } // this should be much longer than keywords.count
let compressed = zip(randomString)
// do assertions
```
that's also closer to the actual payloads where the same keys are usually repeated multiple times.

[the remark this answers is not in the document]
Comment (maxep) [reply Tests/DatadogTests/Datadog/Mocks/SystemFrameworks/FoundationMocks.swift]: >picking a random byte N times, instead of picking a random byte and repeating it N times.

Yes, but that's still fails most of the time. Instead, I've added an object to encode with random properties. This perform a good randomisation with enough redundancy to allow compression.

File: Sources/Datadog/Core/Upload/RequestBuilder.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (maxep) [reply Sources/Datadog/Core/Upload/RequestBuilder.swift:135]: I'v added an internal `warning` when payload fails to compress, LMKWYT!

File: Sources/Datadog/Core/Upload/RequestBuilder.swift [hunk 7]
```diff
@@ -126,9 +127,25 @@ internal struct RequestBuilder {
...
-        request.allHTTPHeaderFields = headers
-        request.httpBody = data
 
+        if let body = Deflate.encode(data) {
+            headers[HTTPHeader.contentEncodingHeaderField] = "deflate"
+            request.httpBody = body
+        } else {
+            request.httpBody = data
+#if DD_SDK_ENABLE_INTERNAL_MONITORING
+            InternalMonitoringFeature.instance?.monitor.sdkLogger.warn(
```
Comment (ncreated) [inline Sources/Datadog/Core/Upload/RequestBuilder.swift:137]: `RequestBuilder` is [also used by the `InternalMonitoringFeature` itself](https://github.com/DataDog/dd-sdk-ios/blob/fcbacd6f7a2cfaaa9159cb6140a5ca19453f492f/Sources/Datadog/InternalMonitoring/InternalMonitoringFeature.swift#L69). To avoid infinite loop in IM (_"send internal log → report compression warning → send internal log → ..."_ we should inject the `monitor: InternalMonitor?` as a dependency to `RequestBuilder` and `nil-ify` it for the `InternalMonitoringFeature`. Similar to [what we do in `DataUploadWorker`](https://github.com/DataDog/dd-sdk-ios/blob/fcbacd6f7a2cfaaa9159cb6140a5ca19453f492f/Sources/Datadog/Core/Upload/DataUploadWorker.swift#L28-L29).

File: Sources/Datadog/Core/Upload/RequestBuilder.swift [hunk 8]
```diff
@@ -126,9 +127,25 @@ internal struct RequestBuilder {
...
+                "Failed to compress request payload",
+                attributes: [
+                    "url": url,
+                    "headers": headers,
+                    "base64-data": data.base64EncodedString()
```
Comment (ncreated) [inline Sources/Datadog/Core/Upload/RequestBuilder.swift:142]: I'm not sure if we want to send `headers` and `base64-data` fields:
* `headers` - Since V2 (https://github.com/DataDog/dd-sdk-ios/pull/562) headers include client-token, so we will be sending foreign tokens to our org.
* `base64-data` - This means sending serialized event, which may include sensitive information, e.g. 1/ we can't send user name & email without requiring extra GDPR clause; e.g. 2/ we don't want to collect secrets from unredacted RUM Resource urls occurring in the payload. 

Unless I'm overlooking some troubleshooting scenario, IMO this could be enough:
* `url` - to know which feature has sent the request (log / trace / rum),
* `uncompressed-size` - to know the uncompressed size (`data.count`) of the request.

This way, if we observe compression warnings for `~100KB` sent from logging feature, that would mean sth actionable. IMO we don't need to know the exact data as we know the schema and size. How does it sound?
